accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1110,9 +1110,8 @@ Assume that _DV_ is an extension type declaration named `Name`, and
 `V1` occurs as one of the `<type>`s in the `<interfaces>` of _DV_. In
 this case we say that `V1` is a _superinterface_ of _DV_.
 
-If _DV_ does not include an `<interfaces>` clause then _DV_ has
-`Object?` or `Object` as a direct superinterface, according to the subtype
-relation which was specified earlier.
+_DV_ has `Object?` or `Object` as a direct superinterface, according to the
+subtype relation which was specified earlier.
```

Comment:
I have at least two worries her.

First the practical: It's potentially a mistake to make `Object` be a direct superinterface.
Example:
```dart
class Bad {
  bool operator ==(covariant Bad other) => something;
}
extension type MoreBad(Bad _) implements Bad {
  /// ...
}
void main() {
  MoreBad(Bad()) == 1; // No error
```
By making `Object` a *direct* superinterface of `MoreBad`, the member signature of `operator==` should be the combined member signatures of `Bad` and `Object`, which is `bool operator==(covariant Object)`. 
That's probably not intended.

(I see the behavior above today if I add `, Object` to the implements list. Adding things to implements lists is not a no-op, even if the type already implements that interface.)

If being a direct superinterface does *not* mean that the member signature should be included in the combined member signature, it needs to be stated very, very clearly. And probably should just not use "direct superinterface" at all.

So, maybe don't include `Object` or `Object?` if not necesary. Something like:
> If _DV_ does not include an `<interfaces>` clause, then _DV_ has `Object` as direct superinterface
> if the declared representation representation type is non-nullable _(a subtype of `Object`)_,
> otherwise it has `Object?` as direct superinterface.
> If _DV_ does include an `<interfaces>` clause, then it has `Object` as direct superinterface if
> the declared representation representation type is non-nullable, and non of the declared 
> superinterfaces from the `<interfaces>` clause have `Object` as superinterface.

That is, we add a direct superinterface *only* if it's needed.

Reply:
It does get more complex, but we could add a case like this:

> No warning is emitted for a downcast from an extension type `E1` to another extension type `E2` in the case where the instantiated representation type of `E1` differs from the instantiated representation type of `E2`.

We also considered the possibility that an extension type can declare that it is 'open' by means of a metadata annotation:

```dart
@openExtensionType
extension type Foo(int it) {
  // For some reason, we don't want to export the interface of int,
  // so we don't have `implements int`. But we _do_ want to allow
  // casting into this extension type: "Absolutely any `int` will do".
  ... // Members.
}

void main() {
  var foo = Foo(42);
  int i = foo; // Compile-time error, not a subtype.
  foo = i as Foo; // OK, no lint, `Foo` is open.
  var xs = [1, 2, 3];
  var ys = xs as List<Foo>; // OK, no lint.
}
```

The lint would treat type tests/casts/matches on open extension types like regular type tests/casts/matches on the representation type.